### PR TITLE
Add a couple integration tests

### LINKED ISSUE #2530: QScript incorrect when we join on distinct
```sql
select * from patients as p join (select distinct state from zips) as z on p.state = z.state
```

In the `QScript` we `ProjectField` of `"state"` in the result of the reducer, and then we `ProjectField` of `"state"` on the result of _that_ in the join condition. We only need one of these projections.

Additionally, on new mongo, the bucket of the reduction is equal to the reduction, yet we are supposed to detect this case and reference the bucket in the reduction instead of duplicating it. 

The `QScript` for `distinct` has changed dramatically on the new mongo branch (compared to master), so unfortunately we probably need to fix this on the new mongo branch.

### LINKED ISSUE #2720: Cannot use `?? {}` in `OBJECT_CONCAT` or `DELETE_FIELD` arguments
Example query that is rejected:

``` sql
SELECT OBJECT_CONCAT(DELETE_FIELD(t0, "b"), t0 ?? {}) FROM `/spark-local/structure_test_deep.json` AS t0
```

Error message from quasar:
``` json
{
  "error": {
    "status": "Error in query.",
    "detail": {
      "message": "Unknown arguments: List(Obj(Map(),Some(Top)), Coproduct(Obj(Map(),Some(Top)),Const(Obj(Map()))))"
    }
  }
}
```

However, using either `t0` or `{}` alone works fine:
``` sql
SELECT OBJECT_CONCAT(DELETE_FIELD(t0, "b"), {}) FROM `/spark-local/structure_test_deep.json` AS t0
SELECT OBJECT_CONCAT(DELETE_FIELD(t0, "b"), t0) FROM `/spark-local/structure_test_deep.json` AS t0
```

The issue seems to be specifically with using constant values with `??`, as using a map on the RHS with some value computed from t0 is fine, but other constant map values are not.

### PR DESCRIPTION
**Pull request by alissapajer**
Closes #2530
Closes #2720

### CODE REVIEW

**Comment by alissapajer** [comment]
@djspiewak the appveyor build was a success, but it hung at the very end. I'm okay merging as is, would you like it restarted?

File: it/src/main/resources/tests/ifundefinedInFunction.test [hunk 1]
```diff
@@ -2,7 +2,7 @@
     "name": "if undefined in function",
 
     "backends": {
-        "couchbase": "ignoreFieldOrder",
+        "couchbase": "pendingIgnoreFieldOrder",
```
Comment (djspiewak) [inline it/src/main/resources/tests/ifundefinedInFunction.test]: Isn't this actually being caused by a mongo bug?

Comment (alissapajer) [reply it/src/main/resources/tests/ifundefinedInFunction.test]: If you meant to comment on the ignore `_id` line, this may be tracked in #3176.

Comment (alissapajer) [reply it/src/main/resources/tests/ifundefinedInFunction.test]: Also, this is an outdated diff. What are you asking?

**Pull request merged**